src/Algebra/Properties/CommutativeMonoid.agda

```diff
@@ -0,0 +1,189 @@
+------------------------------------------------------------------------
+-- The Agda standard library
+--
+-- Some derivable properties
+------------------------------------------------------------------------
+
+open import Algebra
+
+module Algebra.Properties.CommutativeMonoid {g₁ g₂} (M : CommutativeMonoid g₁ g₂) where
+
+open import Algebra.Operations.CommutativeMonoid M
+open import Algebra.CommutativeMonoidSolver M
+open import Relation.Binary as B using (_Preserves_⟶_)
+open import Function
+open import Function.Equality using (_⟨$⟩_)
+open import Data.Product
+open import Data.Nat using (ℕ; zero; suc)
+open import Data.Fin using (Fin; punchIn; zero; suc)
+open import Data.List as List using ([]; _∷_)
+import Data.Fin.Properties as FP
+open import Data.Fin.Permutation as Perm using (Permutation; Permutation′; _⟨$⟩ˡ_; _⟨$⟩ʳ_)
+open import Data.Fin.Permutation.Components as PermC
+open import Data.Table as Table
+open import Data.Table.Relation.Equality as TE using (_≗_)
+open import Data.Unit using (tt)
+open import Data.Empty using (⊥-elim)
+import Data.Table.Properties as TP
+open import Relation.Binary.PropositionalEquality as P using (_≡_)
+open import Relation.Nullary using (yes; no)
+open import Relation.Nullary.Decidable using (⌊_⌋)
+
+open CommutativeMonoid M
+  renaming
+  ( ε to 0#
+  ; _∙_ to _+_
+  ; ∙-cong to +-cong
+  ; identity to +-identity
+  ; assoc to +-assoc
+  ; comm to +-comm
+  )
+open import Algebra.FunctionProperties _≈_
+open import Relation.Binary.EqReasoning setoid
+
+module _ {n} where
+  open B.Setoid (TE.setoid setoid n) public
+    using ()
+    renaming (_≈_ to _≋_)
+
+-- When summing over a function from a finite set, we can pull out any value and move it to the front.
+
+sumₜ-punchIn : ∀ {n} t (i : Fin (suc n)) → sumₜ t ≈ lookup t i + sumₜ (rearrange (punchIn i) t)
+sumₜ-punchIn f zero = refl
+sumₜ-punchIn {zero} t (suc ())
+sumₜ-punchIn {suc n} t (suc i) =
+  begin
+    x + sumₜ (tail t)  ≈⟨ +-cong refl (sumₜ-punchIn (tail t) i) ⟩
+    x + (y + z)        ≈⟨ solve 3 (λ x y z → x ⊕ (y ⊕ z) ⊜ y ⊕ (x ⊕ z)) refl x y z ⟩
+    y + (x + z)        ∎
+  where
+  x = head t
+  y = lookup t (suc i)
+  z = sumₜ (rearrange (punchIn i) (tail t))
+
+-- '_≈_' is a congruence over 'sumTable n'.
+
+sumₜ-cong-≈ : ∀ {n} → sumₜ {n} Preserves _≋_ ⟶ _≈_
+sumₜ-cong-≈ {zero} p = refl
+sumₜ-cong-≈ {suc n} p = +-cong (p _) (sumₜ-cong-≈ (p ∘ suc))
+
+-- '_≡_' is a congruence over 'sum n'.
+
+sumₜ-cong-≡ : ∀ {n} → sumₜ {n} Preserves _≗_ ⟶ _≡_
+sumₜ-cong-≡ {zero} p = P.refl
+sumₜ-cong-≡ {suc n} p = P.cong₂ _+_ (p _) (sumₜ-cong-≡ (p ∘ suc))
+
+-- The sum over the constantly zero function is zero.
+
+sumₜ-zero : ∀ n → sumₜ (replicate {n} 0#) ≈ 0#
+sumₜ-zero (zero) = refl
+sumₜ-zero (suc n) =
+  begin
```

Comment:
We usually put `begin` on the previous line in the library, and then you can de-indent the actual proof by 2 spaces which gives you more room to play with which can be important!

Reply:
What would you do with the `where` clause if `begin` is on the previous line? This looks weird to me:
```agda
sumₜ-punchIn {suc n} t (suc i) = begin
  x + sumₜ (tail t)  ≈⟨ +-cong refl (sumₜ-punchIn (tail t) i) ⟩
  x + (y + z)        ≈⟨ solve 3 (λ x y z → x ⊕ (y ⊕ z) ⊜ y ⊕ (x ⊕ z)) refl x y z ⟩
  y + (x + z)        ∎
  where
  x = head t
  y = lookup t (suc i)
  z = sumₜ (rearrange (punchIn i) (tail t))
```

and this is inconsistent with the indentation after `where` in other places:
```agda
sumₜ-punchIn {suc n} t (suc i) = begin
  x + sumₜ (tail t)  ≈⟨ +-cong refl (sumₜ-punchIn (tail t) i) ⟩
  x + (y + z)        ≈⟨ solve 3 (λ x y z → x ⊕ (y ⊕ z) ⊜ y ⊕ (x ⊕ z)) refl x y z ⟩
  y + (x + z)        ∎
  where
    x = head t
    y = lookup t (suc i)
    z = sumₜ (rearrange (punchIn i) (tail t))
```

Reply:
I don't think there's anything wrong with the first one? The `where` clearly separates the two? 

Also I think the proof would clearer if you gave `x` `y` and `z` more meaningful names, something like `t_0`, `t_i` and `\sigma[t-t_i]`?

Reply:
Also it's just occurred to me that we should define a function `remove` in `Data.Table` and prove some properties about it. That would massively clean up these proofs right?

Reply:
> I don't think there's anything wrong with the first one? The where clearly separates the two.

I adopted the habit of starting begin on the next line specifically to avoid this in my own code! I just don't think text (i.e. the word `where`) is a very good visual separator on its own.

> Also it's just occurred to me that we should define a function remove in Data.Table and prove some properties about it. That would massively clean up these proofs right?

I had the same thought about defining that function, but I'm not sure what it would clean up other than making the names clearer (because it's not immediately obvious what `punchIn` does to a table).